### ISSUE TITLE
twisted.internet.test.test_socket fails on Windows Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8749]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)| @rodrigc reported|
|-|-|
|Trac ID|trac#8749|
|Type|enhancement|
|Created|2016-08-12 08:20:06Z|
|Branch|https://github.com/twisted/twisted/tree/8749-rodrigc-socket-py3|

On Windows, with Python 3.5, if I do:

```
python -m twisted.trial twisted.internet.test.test_socket
```

I get these errors:

```
twisted.internet.test.test_socket
  AdoptDatagramPortErrorsTestsBuilder_IOCPReactorTests
    test_invalidAddressFamily ...                                     [SKIPPED]
    test_invalidDescriptor ...                                        [SKIPPED]
    test_stopOnlyCloses ...                                           [SKIPPED]
  AdoptDatagramPortErrorsTestsBuilder_PortableGIReactorTests
    test_invalidAddressFamily ...                                     [SKIPPED]
    test_invalidDescriptor ...                                        [SKIPPED]
    test_stopOnlyCloses ...                                           [SKIPPED]
  AdoptDatagramPortErrorsTestsBuilder_PortableGtk3ReactorTests
    test_invalidAddressFamily ...                                     [SKIPPED]
    test_invalidDescriptor ...                                        [SKIPPED]
    test_stopOnlyCloses ...                                           [SKIPPED]
  AdoptDatagramPortErrorsTestsBuilder_PortableGtkReactorTests
    test_invalidAddressFamily ...                                     [SKIPPED]
    test_invalidDescriptor ...                                        [SKIPPED]
    test_stopOnlyCloses ...                                           [SKIPPED]
  AdoptDatagramPortErrorsTestsBuilder_SelectReactorTests
    test_invalidAddressFamily ...                                          [OK]
    test_invalidDescriptor ...                                           [FAIL]
    test_stopOnlyCloses ...                                             [ERROR]
  AdoptDatagramPortErrorsTestsBuilder_Win32ReactorTests
    test_invalidAddressFamily ...                                          [OK]
    test_invalidDescriptor ...                                           [FAIL]
    test_stopOnlyCloses ...                                             [ERROR]
  AdoptStreamConnectionErrorsTestsBuilder_IOCPReactorTests
    test_invalidAddressFamily ...                                     [SKIPPED]
  AdoptStreamConnectionErrorsTestsBuilder_PortableGIReactorTests
    test_invalidAddressFamily ...                                     [SKIPPED]
  AdoptStreamConnectionErrorsTestsBuilder_PortableGtk3ReactorTests
    test_invalidAddressFamily ...                                     [SKIPPED]
  AdoptStreamConnectionErrorsTestsBuilder_PortableGtkReactorTests
    test_invalidAddressFamily ...                                     [SKIPPED]
  AdoptStreamConnectionErrorsTestsBuilder_SelectReactorTests
    test_invalidAddressFamily ...                                          [OK]
  AdoptStreamConnectionErrorsTestsBuilder_Win32ReactorTests
    test_invalidAddressFamily ...                                          [OK]
  AdoptStreamPortErrorsTestsBuilder_IOCPReactorTests
    test_invalidAddressFamily ...                                     [SKIPPED]
    test_invalidDescriptor ...                                        [SKIPPED]
    test_stopOnlyCloses ...                                           [SKIPPED]
  AdoptStreamPortErrorsTestsBuilder_PortableGIReactorTests
    test_invalidAddressFamily ...                                     [SKIPPED]
    test_invalidDescriptor ...                                        [SKIPPED]
    test_stopOnlyCloses ...                                           [SKIPPED]
  AdoptStreamPortErrorsTestsBuilder_PortableGtk3ReactorTests
    test_invalidAddressFamily ...                                     [SKIPPED]
    test_invalidDescriptor ...                                        [SKIPPED]
    test_stopOnlyCloses ...                                           [SKIPPED]
  AdoptStreamPortErrorsTestsBuilder_PortableGtkReactorTests
    test_invalidAddressFamily ...                                     [SKIPPED]
    test_invalidDescriptor ...                                        [SKIPPED]
    test_stopOnlyCloses ...                                           [SKIPPED]
  AdoptStreamPortErrorsTestsBuilder_SelectReactorTests
    test_invalidAddressFamily ...                                       [ERROR]
    test_invalidDescriptor ...                                           [FAIL]
    test_stopOnlyCloses ...                                             [ERROR]
  AdoptStreamPortErrorsTestsBuilder_Win32ReactorTests
    test_invalidAddressFamily ...                                       [ERROR]
    test_invalidDescriptor ...                                           [FAIL]
    test_stopOnlyCloses ...                                             [ERROR]
  IReactorSocketVerificationTestsBuilder_IOCPReactorTests
    test_provider ...                                                 [SKIPPED]
  IReactorSocketVerificationTestsBuilder_PortableGIReactorTests
    test_provider ...                                                 [SKIPPED]
  IReactorSocketVerificationTestsBuilder_PortableGtk3ReactorTests
    test_provider ...                                                 [SKIPPED]
  IReactorSocketVerificationTestsBuilder_PortableGtkReactorTests
    test_provider ...                                                 [SKIPPED]
  IReactorSocketVerificationTestsBuilder_SelectReactorTests
    test_provider ...                                                      [OK]
  IReactorSocketVerificationTestsBuilder_Win32ReactorTests
    test_provider ...                                                      [OK]

===============================================================================
[SKIPPED]
twisted.internet.iocpreactor.reactor.IOCPReactor does not provide IReactorSocket

twisted.internet.test.test_socket.AdoptDatagramPortErrorsTestsBuilder_IOCPReactorTests.test_invalidAddressFamily
twisted.internet.test.test_socket.AdoptDatagramPortErrorsTestsBuilder_IOCPReactorTests.test_invalidDescriptor
twisted.internet.test.test_socket.AdoptDatagramPortErrorsTestsBuilder_IOCPReactorTests.test_stopOnlyCloses
twisted.internet.test.test_socket.AdoptStreamConnectionErrorsTestsBuilder_IOCPReactorTests.test_invalidAddressFamily
twisted.internet.test.test_socket.AdoptStreamPortErrorsTestsBuilder_IOCPReactorTests.test_invalidAddressFamily
twisted.internet.test.test_socket.AdoptStreamPortErrorsTestsBuilder_IOCPReactorTests.test_invalidDescriptor
twisted.internet.test.test_socket.AdoptStreamPortErrorsTestsBuilder_IOCPReactorTests.test_stopOnlyCloses
twisted.internet.test.test_socket.IReactorSocketVerificationTestsBuilder_IOCPReactorTests.test_provider
===============================================================================
[SKIPPED]
No module named 'gi'

twisted.internet.test.test_socket.AdoptDatagramPortErrorsTestsBuilder_PortableGIReactorTests.test_invalidAddressFamily
twisted.internet.test.test_socket.AdoptDatagramPortErrorsTestsBuilder_PortableGIReactorTests.test_invalidDescriptor
twisted.internet.test.test_socket.AdoptDatagramPortErrorsTestsBuilder_PortableGIReactorTests.test_stopOnlyCloses
twisted.internet.test.test_socket.AdoptDatagramPortErrorsTestsBuilder_PortableGtk3ReactorTests.test_invalidAddressFamily
twisted.internet.test.test_socket.AdoptDatagramPortErrorsTestsBuilder_PortableGtk3ReactorTests.test_invalidDescriptor
twisted.internet.test.test_socket.AdoptDatagramPortErrorsTestsBuilder_PortableGtk3ReactorTests.test_stopOnlyCloses
twisted.internet.test.test_socket.AdoptStreamConnectionErrorsTestsBuilder_PortableGIReactorTests.test_invalidAddressFamily
twisted.internet.test.test_socket.AdoptStreamConnectionErrorsTestsBuilder_PortableGtk3ReactorTests.test_invalidAddressFamily
twisted.internet.test.test_socket.AdoptStreamPortErrorsTestsBuilder_PortableGIReactorTests.test_invalidAddressFamily
twisted.internet.test.test_socket.AdoptStreamPortErrorsTestsBuilder_PortableGIReactorTests.test_invalidDescriptor
twisted.internet.test.test_socket.AdoptStreamPortErrorsTestsBuilder_PortableGIReactorTests.test_stopOnlyCloses
twisted.internet.test.test_socket.AdoptStreamPortErrorsTestsBuilder_PortableGtk3ReactorTests.test_invalidAddressFamily
twisted.internet.test.test_socket.AdoptStreamPortErrorsTestsBuilder_PortableGtk3ReactorTests.test_invalidDescriptor
twisted.internet.test.test_socket.AdoptStreamPortErrorsTestsBuilder_PortableGtk3ReactorTests.test_stopOnlyCloses
twisted.internet.test.test_socket.IReactorSocketVerificationTestsBuilder_PortableGIReactorTests.test_provider
twisted.internet.test.test_socket.IReactorSocketVerificationTestsBuilder_PortableGtk3ReactorTests.test_provider
===============================================================================
[SKIPPED]
module 'twisted.internet' has no attribute 'gtk2reactor'

twisted.internet.test.test_socket.AdoptDatagramPortErrorsTestsBuilder_PortableGtkReactorTests.test_invalidAddressFamily
twisted.internet.test.test_socket.AdoptDatagramPortErrorsTestsBuilder_PortableGtkReactorTests.test_invalidDescriptor
twisted.internet.test.test_socket.AdoptDatagramPortErrorsTestsBuilder_PortableGtkReactorTests.test_stopOnlyCloses
twisted.internet.test.test_socket.AdoptStreamConnectionErrorsTestsBuilder_PortableGtkReactorTests.test_invalidAddressFamily
twisted.internet.test.test_socket.AdoptStreamPortErrorsTestsBuilder_PortableGtkReactorTests.test_invalidAddressFamily
twisted.internet.test.test_socket.AdoptStreamPortErrorsTestsBuilder_PortableGtkReactorTests.test_invalidDescriptor
twisted.internet.test.test_socket.AdoptStreamPortErrorsTestsBuilder_PortableGtkReactorTests.test_stopOnlyCloses
twisted.internet.test.test_socket.IReactorSocketVerificationTestsBuilder_PortableGtkReactorTests.test_provider
===============================================================================
[FAIL]
Traceback (most recent call last):
  File "c:\Users\Craig\twisted3\build\py35-tests-windows\lib\site-packages\twisted\internet\test\test_socket.py", line 196, in test_invalidDescriptor
    self.assertEqual(exc.args[0], errno.EBADF)
  File "c:\Users\Craig\twisted3\build\py35-tests-windows\lib\site-packages\twisted\trial\_synctest.py", line 425, in assertEqual
    super(_Assertions, self).assertEqual(first, second, msg)
  File "c:\python35\Lib\unittest\case.py", line 820, in assertEqual
    assertion_func(first, second, msg=msg)
  File "c:\python35\Lib\unittest\case.py", line 813, in _baseAssertEqual
    raise self.failureException(msg)
twisted.trial.unittest.FailTest: 10038 != 9

twisted.internet.test.test_socket.AdoptDatagramPortErrorsTestsBuilder_SelectReactorTests.test_invalidDescriptor
twisted.internet.test.test_socket.AdoptDatagramPortErrorsTestsBuilder_Win32ReactorTests.test_invalidDescriptor
===============================================================================
[FAIL]
Traceback (most recent call last):
  File "c:\Users\Craig\twisted3\build\py35-tests-windows\lib\site-packages\twisted\internet\test\test_socket.py", line 82, in test_invalidDescriptor
    self.assertEqual(exc.args[0], errno.EBADF)
  File "c:\Users\Craig\twisted3\build\py35-tests-windows\lib\site-packages\twisted\trial\_synctest.py", line 425, in assertEqual
    super(_Assertions, self).assertEqual(first, second, msg)
  File "c:\python35\Lib\unittest\case.py", line 820, in assertEqual
    assertion_func(first, second, msg=msg)
  File "c:\python35\Lib\unittest\case.py", line 813, in _baseAssertEqual
    raise self.failureException(msg)
twisted.trial.unittest.FailTest: 10038 != 9

twisted.internet.test.test_socket.AdoptStreamPortErrorsTestsBuilder_SelectReactorTests.test_invalidDescriptor
twisted.internet.test.test_socket.AdoptStreamPortErrorsTestsBuilder_Win32ReactorTests.test_invalidDescriptor
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "c:\Users\Craig\twisted3\build\py35-tests-windows\lib\site-packages\twisted\internet\test\test_socket.py", line 235, in test_stopOnlyCloses
    portSocket.fileno(), portSocket.family, DatagramProtocol())
  File "c:\Users\Craig\twisted3\build\py35-tests-windows\lib\site-packages\twisted\internet\posixbase.py", line 468, in adoptDatagramPort
    maxPacketSize=maxPacketSize)
  File "c:\Users\Craig\twisted3\build\py35-tests-windows\lib\site-packages\twisted\internet\udp.py", line 151, in _fromListeningDescriptor
    interface = port.getsockname()[0]
builtins.OSError: [WinError 10022] An invalid argument was supplied

twisted.internet.test.test_socket.AdoptDatagramPortErrorsTestsBuilder_SelectReactorTests.test_stopOnlyCloses
twisted.internet.test.test_socket.AdoptDatagramPortErrorsTestsBuilder_Win32ReactorTests.test_stopOnlyCloses
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "c:\Users\Craig\twisted3\build\py35-tests-windows\lib\site-packages\twisted\internet\test\test_socket.py", line 94, in test_invalidAddressFamily
    port.listen(1)
builtins.OSError: [WinError 10022] An invalid argument was supplied

twisted.internet.test.test_socket.AdoptStreamPortErrorsTestsBuilder_SelectReactorTests.test_invalidAddressFamily
twisted.internet.test.test_socket.AdoptStreamPortErrorsTestsBuilder_Win32ReactorTests.test_invalidAddressFamily
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "c:\Users\Craig\twisted3\build\py35-tests-windows\lib\site-packages\twisted\internet\test\test_socket.py", line 117, in test_stopOnlyCloses
    portSocket.listen(1)
builtins.OSError: [WinError 10022] An invalid argument was supplied

twisted.internet.test.test_socket.AdoptStreamPortErrorsTestsBuilder_SelectReactorTests.test_stopOnlyCloses
twisted.internet.test.test_socket.AdoptStreamPortErrorsTestsBuilder_Win32ReactorTests.test_stopOnlyCloses
-------------------------------------------------------------------------------
Ran 48 tests in 0.122s

FAILED (skips=32, failures=4, errors=6, successes=6)
```

<details><summary>Searchable metadata</summary>

```
trac-id__8749 8749
type__enhancement enhancement
reporter__rodrigc rodrigc
priority__normal normal
milestone__Python_3_x Python-3.x
branch__8749_rodrigc_socket_py3 8749-rodrigc-socket-py3
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1470990006409679 1470990006409679
changetime__1471159029000375 1471159029000375
version__None None
owner__GitHub__noreply_____ GitHub <noreply@...>

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_1"></a>|@rodrigc commented|
|-|-|

https://github.com/twisted/twisted/pull/490

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_2"></a>|@markrwilliams commented|
|-|-|

Some minor comments on the PR.

Also, it appears that the windows7-64-py2 failed on unrelated tests: https://buildbot.twistedmatrix.com/builders/windows7-64-py2.7/builds/1389

Please address the comment on the if statement in the PR, then merge.

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|GitHub set owner to GitHub<br>GitHub set status to `closed`|
|-|-|

In changeset d3c7671c4fdcf4d3e52d5e952795155aad75b6ae
```
#!CommitTicketReference repository="" revision="d3c7671c4fdcf4d3e52d5e952795155aad75b6ae"
Merge pull request #490 from twisted/8749-rodrigc-socket-py3

Author: hawkowl, rodrigc
Reviewer: markrwilliams
Fixes: #8749

Fix test_tcp on Windows Python 3
```
